lib/scholar/cluster/gmm.ex

```diff
@@ -0,0 +1,433 @@
+          if max_lower_bound < lower_bound do
+            {lower_bound, params}
+          else
+            {max_lower_bound, best_params}
+          end
+
+        {
+          best_params,
+          {run + 1, max_lower_bound, x, key, max_iter, tol, covariance_regularization_eps}
+        }
+      end
+
+    {weights, means, covariances, precisions_cholesky} = best_params
+
+    %__MODULE__{
+      weights: weights,
+      means: means,
+      covariances: covariances,
+      precisions_cholesky: precisions_cholesky
+    }
+  end
+
+  defnp init_params(key, x, opts) do
+    type = Nx.type(x)
+    {num_samples, _} = Nx.shape(x)
+    num_gaussians = opts[:num_gaussians]
+    covariance_regularization_eps = opts[:covariance_regularization_eps]
+
+    k_means_model =
+      Scholar.Cluster.KMeans.fit(x, num_clusters: num_gaussians, num_runs: 1, key: key)
+
+    labels = k_means_model.labels
+    responsibilities = Nx.take(Nx.eye(num_gaussians, type: type), labels)
+
+    {weights, means, covariances, precisions_cholesky} =
+      estimate_gaussian_parameters(x, responsibilities, covariance_regularization_eps)
+
+    weights = weights / num_samples
+    {weights, means, covariances, precisions_cholesky}
+  end
+
+  defnp e_step(x, weights, means, precisions_cholesky) do
+    {log_prob_norm, log_responsibilities} =
+      estimate_log_prob_responsibilities(x, weights, means, precisions_cholesky)
+
+    {Nx.mean(log_prob_norm), log_responsibilities}
+  end
+
+  defnp m_step(x, log_responsibilities, covariance_regularization_eps) do
+    {weights, means, covariances, precisions_cholesky} =
+      estimate_gaussian_parameters(x, Nx.exp(log_responsibilities), covariance_regularization_eps)
+
+    weights = weights / Nx.sum(weights)
+    {weights, means, covariances, precisions_cholesky}
+  end
+
+  defnp estimate_gaussian_parameters(x, responsibilities, covariance_regularization_eps) do
+    nk = Nx.sum(responsibilities, axes: [0])
+    means = Nx.dot(Nx.transpose(responsibilities), x / nk)
+
+    covariances =
+      estimate_gaussian_covariances(x, responsibilities, nk, means, covariance_regularization_eps)
+
+    precisions_cholesky = compute_precisions_cholesky(covariances)
+    {nk, means, covariances, precisions_cholesky}
+  end
+
+  defnp estimate_gaussian_covariances(
+          x,
+          responsibilities,
+          nk,
+          means,
+          covariance_regularization_eps
+        ) do
+    type = Nx.type(x)
+    {num_gaussians, num_features} = Nx.shape(means)
+
+    {covariances, _} =
+      while {
+              covariances =
+                Nx.tensor(0.0, type: type)
+                |> Nx.broadcast({num_gaussians, num_features, num_features}),
+              {k = 0, x, responsibilities, nk, means, covariance_regularization_eps}
+            },
+            k < num_gaussians do
+        diff = x - means[k]
+        covariance = Nx.dot(responsibilities[[.., k]] * Nx.transpose(diff), diff) / nk[k]
```

Comment:
```suggestion
        covariance = Nx.dot(responsibilities[[.., k]] * diff, [-2], diff / nk[k], [-2])
```

Reply:
This won't work as `Nx.transpose` is only applied to `diff`, not the entire product. Adding another axis to `responsibilities[[.., k]]` solves the problem.
`covariance = Nx.dot(diff * Nx.new_axis(responsibilities[[.., k]], 1), [0], diff / nk[k], [0])`